compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1998,8 +2000,24 @@ public void visit(BLangInvocation invocationExpr) {
         }
 
         if (invocationExpr.actionInvocation || invocationExpr.async) {
-            validateActionInvocation(invocationExpr.pos, invocationExpr);
+            if (invocationExpr.async && this.withinLockBlock) {
+                logAsyncCallWithinLockError(invocationExpr);
+            } else {
+                validateActionInvocation(invocationExpr.pos, invocationExpr);
+            }
```

Comment:
Shall we switch the checks and return early?
```suggestion
            if (!invocationExpr.async || !this.withinLockBlock) {
                validateActionInvocation(invocationExpr.pos, invocationExpr);
                return;
            }
            // log the error here
```

Reply:
Fixed